index.rst

```diff
@@ -13,17 +13,256 @@
+We could use that information with the new ``daf_relation`` classes to easily integrate them with the query system, allowing user queries to not just test for Datastore existence, but query on and report Datastore specific-fields like file size.
+We'd also of course provide a way for users to inspect which such fields are available, since Datastore record fields can change from implementation to implementation.
+
+Datastore methods that add new datasets to the repository would be modified to return a collection of records describing those datasets, again for Butler to pass to Registry.
+Datastore methods that read datasets or interpret the records describing them would be modified to accept those records from Butler (which fetches them from Registry).
+Some Datastore existence-check methods would go away entirely (e.g. ``knows``), as their functionality is subsumed by Registry dataset queries, while others would change their behavior to checking for artifact existence *given* records.
+``Registry.insertDatasets`` would be modified to accept datastore records for storage, and ``Registry.findDataset`` would be modified to return dataset records as well as a ``DatasetRef``.
+
+.. note::
+   All ``DatasetRef`` objects in this technote are assumed to be resolved; unresolved ``DatasetRef`` objects are already slated to go away per :jira:`RFC-888`.
+
+This proposal formalizes what we are already doing during no-database batch execution, while taking advantage of new developments - UUIDs and ``daf_relation`` - to simplify the Registry/Datastore boundary.
+It would involve considerable code changes, but more removals than additions, and the vast majority of these would be behind the scenes or of minimal impact to users (e.g. ``Butler.datastore`` and ``Registry.insertDatasets`` are not formally private, but they should be, and are already widely recognized as for internal use only).
+
+Implementation of important butler operations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In order for this proposal to be considered viable, we need to look carefully at how we would implement joint Registry/Datastore operations that we want to maintain consistency.
+Note that later sections in this technical note will expand upon these and ultimately take the form of prototype Python code, as further changes are added to the proposal.
+
+``Butler.put``
+""""""""""""""
+
+1. Obtain a valid expanded data ID from the Registry.
+   In the vast majority of cases (i.e. QuantumGraph execution) this done well in advance of the actual ``put`` call.
+
+2. Construct a ``DatasetRef`` by generating an appropriate UUID and using an existing or soon-to-exist RUN name.
+   This will also typically occur well before the rest of the ``put``, as part of QuantumGraph generation.
```

Comment:
As discussed in DMTN-242, do we want an initial ping to registry asking whether this new DatasetRef is already in conflict with registry? Failing early when we know it will fail later seems like a good idea. Obviously in a race the absence of a registry entry does not mean there won't be one when the registry update happens at the end but it will at least potentially save us the bother of doing the slow serializations. This check would be very fast for QBB scenario and for client server will cause a small delay.